lib/nerve/rate_limiter.rb

```diff
@@ -0,0 +1,45 @@
+module Nerve
+  # RateLimiter implements a TokenBucket-based rate-limiting strategy.
+  # It allows an average of `average_rate` tokens per `period`, with a
+  # maximum burst of `max_burst` when the bucket is full.
+  # `period` is the time period for `average_rate`, in seconds.
+  # Note: a single instance of RateLimiter is *not* thread-safe.
+  # See: https://en.wikipedia.org/wiki/Token_bucket
+  class RateLimiter
+    def initialize(average_rate: 1, max_burst: 10, period: 1)
+      raise TypeError, "average_rate should be numeric" unless average_rate.is_a? Numeric
+      raise TypeError, "max_burst should be numeric" unless max_burst.is_a? Numeric
+      raise TypeError, "period should be numeric" unless period.is_a? Numeric
+
+      @average_rate = average_rate
+      @max_burst = max_burst
+      @period = period
+
+      @tokens = @average_rate
+      @last_refill = Time.now
+    end
+
+    # Consume a token, if one is available. Returns true if a token was consumed
+    # and false otherwise. For a rate-limited action, the action should be
+    # executed only when `consume` returns true.
+    def consume
+      refill_tokens
+      return false unless @tokens >= 1
+
+      @tokens -= 1
+      return true
+    end
+
+    private
+
+    def refill_tokens
+      now = Time.now
+      elapsed = (now - @last_refill).to_f / @period
+      delta_tokens = (@average_rate * elapsed).floor
+      return nil unless delta_tokens >= 1
```

Comment:
this could introduce some level of inaccuracy, delta_tokens can be 0.999999

Reply:
Because of floating-point arithmetic? That's a good point, though that would be resolved of sorts on the next call to `consume`. I can round off `delta_tokens` instead of taking the floor, but then we want the precision to be based off of the average rate (i.e. a precision of 0.1 is fine if you allow 10000 requests/second, but not if you allow 0.01 requests/second).

That's easy enough to implement, but is the added complexity worthwhile? I'm not sure if the inaccuracy will be relevant frequently enough (especially since we are the only users of the rate limiter) to justify it.

Reply:
I understand that we don't need high precision in current use case, since state transition triggered by health check will most likely not be at high frequency. However, we never know someone later may use it for other purpose later in the project. It's better to not make assumption.

Here if we just remove the `floor` and `return nil unless >= 1` check, then it should work for high precision as well.

Reply:
Okay, but doesn't that mean that floating-point error (from incrementing `@tokens`) adds up over time, and so `@tokens` will have increasing amounts of error over long runs?

Not likely to be an issue for our use case, but I don't think this solves the issue from the perspective of a general rate limiter

Reply:
can you be more specific (maybe sharing an example) about concern that  floating-point error increasing over long runs?

Reply:
We calculate the number of tokens to add as

```
delta_tokens = @average_rate * elapsed
```

and then add it in as (assuming the `max_burst` is high enough)

```
@tokens += delta
```

That means that any floating-point error from the multiplication (or the addition, because the results can be stored inaccurately as 0.100001, for example) will just add up over time. Of course it takes a large number of operations for that to have any effect, but we were considering the case of high-frequency operations.

My point is that keeping tokens as a float doesn't resolve the inaccuracy issue, as far as I can tell. Maybe the effect of the floating-point error is very minute and so we can consider it not important, but I wanted to point it out

Reply:
I under the concern now, there is one way to avoid that,  we can store tokens in a temporary variable, and only update member variable (last time as well) after the ">=1" check, that should minimize the accumulation of floating-point error over time.

I am more worried about rounding error, since floating-point error is tiny comparing to that, and take many many rounds of calculation to accumulate.